### PR TITLE
Add chunked read ahead iterator

### PR DESCRIPTION
**Pull request by NatPRoach**
Adds a chunked read ahead iterator, an iterator that asynchronous to the main thread reads ahead and stores the resulting value in a channel, allowing the main thread to grab elements off that channel faster.

Also adds a trait, `IntoChunkedReadAheadIterator`, which is implemented for all iterators. (Or at least all `'static` `Send` iterators)

### CODE REVIEW

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    #[case(16)] // larger than the inner iterator
+    #[test]
+    fn test_dropping_half_used_iterator_doesnt_explode(#[case] _chunk_size: usize) {
+        let test_vec = vec![0usize, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9];
+
+        let mut chunked_iter = test_vec.iter(); // TODO add into chunked readahead
+        for _ in 0..4 {
+            chunked_iter.next();
+        }
+        drop(chunked_iter);
```
Comment (nh13) [inline src/iterators/chunked_read_ahead_iterator.rs]: do you want to vary the _time_ you want to drop it so that we test dropping before it consumes, while it consumes, and after it is all done consuming the original iterator?

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        let mut chunked_iter = test_vec2.iter(); // TODO add into chunked readahead
+
+        loop {
+            let i = regular_iter.next();
+            let j = chunked_iter.next();
+            assert_eq!(i, j);
+            if i.is_none() {
+                break;
+            }
+        }
```
Comment (nh13) [inline src/iterators/chunked_read_ahead_iterator.rs]: want to check that both are empty here?

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        for _ in 0..FAIL_POINT {
+            // Need to pass ownership back and forth to avoid the borrow checker complaining
+            test_iter = panic::catch_unwind(|| {
+                test_iter.next();
+                test_iter
+            })
+            .expect("different error message");
+        }
+        test_iter.next();
+    }
```
Comment (nh13) [inline src/iterators/chunked_read_ahead_iterator.rs]: What should we do when the underlying iterator is exhausted but it panics?  For example, reading lines from a file but the close on the reader fails after returning all the lines?

**Review by nh13 — COMMENTED** [review]
A few more unit tests:
1. handles wrapping an empty iterator
2. handles a very large iterator, where the chunk size is much smaller than the number of items returned
3. after the iterator is exhausted, calling `iter.next()` multiple times should always return none (see this [test case](https://github.com/sourcefrog/readahead-iterator/blob/86d13ea7facfd6a320accdb6439b65f3bec3852a/tests/iterate.rs#L28))
4. a test case where the bound on the channel is very small, such that the inner thread has to block on sending messages, just to make sure that it resumes sending when the channel has space

@tfenne thoughts?

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+                                panic!("{}", extract_info_from_catch_unwind_panic(e));
+                            }
+                        }
+                    }
+                    if chunk.is_empty() {
+                        break;
+                    }
+                    let send_result = sender.send(Some(chunk));
+                    if send_result.is_err() {
+                        // TODO - Logging?
```
Comment (NatPRoach) [inline src/iterators/chunked_read_ahead_iterator.rs]: Wasn't sure if we wanted to log this if errors occurred here. Would be easy to do, but wanted your feedback.

File: src/iterators/chunked_read_ahead_iterator.rs (hunk 3 above)
Comment (NatPRoach) [reply src/iterators/chunked_read_ahead_iterator.rs]: I think that will be covered by the current implementation. I can add a test for that to be sure? would just need to think about how to go about doing so.

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 5]
```diff
@@ -0,0 +1,393 @@
+use std::any::Any;
```
Comment (tfenne) [inline src/iterators/chunked_read_ahead_iterator.rs]: Before we merge we should just talk about package/module structure.  I tend to think that maybe we should just have `src/iter.rs` that contains this code so that the usage looks more like:

```rust
use fgoxide::iter::ChunkedReadAheadIterator
```

But let's do anything like this at the very end and not until the PR is otherwise ready to go.

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 6]
```diff
@@ -0,0 +1,393 @@
...
+use std::panic::{catch_unwind, AssertUnwindSafe};
+use std::sync::mpsc::{sync_channel, Receiver};
+use std::thread::{self, JoinHandle};
+use std::vec::IntoIter;
+
+/// Takes the type returned by a catch unwind block and attempts to convert it into a string.
+/// If it fails at this returns "Unknown Source of Error", otherwise returns the String it
+/// extracted.
+fn extract_info_from_catch_unwind_panic(e: Box<dyn Any + Send>) -> String {
```
Comment (tfenne) [inline src/iterators/chunked_read_ahead_iterator.rs]: Does this belong in this module or somewhere more general?
Perhaps rename to something like `panic_payload_to_string()`?

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 7]
```diff
@@ -0,0 +1,393 @@
...
+                _ => "Unknown Source of Error",
+            };
+            s.to_owned()
+        }
+    }
+}
+
+/// Iterator extension that spawns an additional thread to read-ahead in the iterator. Sends
+/// results back to this object via a channel and returns them in the same manner as a normal
+/// iterator
```
Comment (tfenne) [inline src/iterators/chunked_read_ahead_iterator.rs]: We'll want to expand this significantly before merge, explaining why you'd use this, with examples for how to create etc.

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 8]
```diff
@@ -0,0 +1,393 @@
...
+pub struct ChunkedReadAheadIterator<T: Send + 'static> {
+    /// The recieving object that recieves chunks of ``T``. TODO - make this a Vec<T> when adding
+    /// chunking.
+    receiver: Option<Receiver<Option<Vec<T>>>>,
```
Comment (tfenne) [inline src/iterators/chunked_read_ahead_iterator.rs]: I get why the outer `Option`, so that you can drop it first.  But why is the inner one `Option` also?

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 9]
```diff
@@ -0,0 +1,393 @@
...
+    /// The handle to the thread that was spawned to read ahead on the iterator.
+    join_handle: Option<JoinHandle<()>>,
```
Comment (tfenne) [inline src/iterators/chunked_read_ahead_iterator.rs]: Does this one need to be an option?

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 10]
```diff
@@ -0,0 +1,393 @@
...
+    /// the chunk
+    current_chunk: IntoIter<T>,
+}
+
+impl<T> ChunkedReadAheadIterator<T>
+where
+    T: Send + 'static,
+{
+    /// Creates a new ``Self`` from an existing iterator, and parameters concerning the size of
+    /// underlying buffer elements.
```
Comment (tfenne) [inline src/iterators/chunked_read_ahead_iterator.rs]:
```suggestion
    /// Creates a new ``Self`` from an existing iterator.  Takes two parameters that
    /// control the number of items stored in the read-ahead buffer.  `chunk_size`
    /// refers to how many items are transferred at a time from read-ahead thread and
    /// `chunk_count` controls how many chunks are read ahead.
```

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 11]
```diff
@@ -0,0 +1,393 @@
...
+    ///
+    /// # Panics
+    ///
+    /// - panics if the underlying iterator panics
+    /// - panics if sending final chunk after an internal panic fails
+    /// - panics if sending the None kill signal to the recieiver after an internal panic fails
+    /// - panics if the spawned thread fails to spawn
```
Comment (tfenne) [inline src/iterators/chunked_read_ahead_iterator.rs]: These should be why the `::new()` function panics, not why the returned struct might panic during it's lifetime.  I think the first three should be on the struct, not the new function.

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 12]
```diff
@@ -0,0 +1,393 @@
...
+                            Ok(Some(val)) => chunk.push(val),
+                            Ok(None) => break,
+                            Err(e) => {
+                                sender
+                                    .send(Some(chunk))
+                                    .expect("Error sending final chunk before internal panic");
+                                sender
+                                    .send(None)
+                                    .expect("Error sending final None before internal panic");
+                                panic!("{}", extract_info_from_catch_unwind_panic(e));
```
Comment (tfenne) [inline src/iterators/chunked_read_ahead_iterator.rs]: Instead of doing this I think you should instead do `std::panic::resume_unwind()` which will take `e` directly without you having to "extract" anything from it, _and_ re-raise the panic without re-triggering the panic hooks.

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 13]
```diff
@@ -0,0 +1,393 @@
...
+                    for _ in 0..chunk_size {
+                        match catch_unwind(AssertUnwindSafe(|| inner.by_ref().next())) {
+                            Ok(Some(val)) => chunk.push(val),
+                            Ok(None) => break,
+                            Err(e) => {
+                                sender
+                                    .send(Some(chunk))
+                                    .expect("Error sending final chunk before internal panic");
+                                sender
+                                    .send(None)
```
Comment (tfenne) [inline src/iterators/chunked_read_ahead_iterator.rs]: Rather than using `Option<Vec<T>>` could you not just send a `Vec<T>` and then drop the sender when it's done?

File: src/iterators/chunked_read_ahead_iterator.rs (hunk 4 above)
Comment (tfenne) [reply src/iterators/chunked_read_ahead_iterator.rs]: Nope.  Not really an error right?  The docs are clear that SendErrors should only occur if the channel has been closed, in which case it's not an error.

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 14]
```diff
@@ -0,0 +1,393 @@
...
+                        }
+                    }
+                    if chunk.is_empty() {
+                        break;
+                    }
+                    let send_result = sender.send(Some(chunk));
+                    if send_result.is_err() {
+                        // TODO - Logging?
+                        break;
+                    }
```
Comment (tfenne) [inline src/iterators/chunked_read_ahead_iterator.rs]:
```suggestion
                    // If there is nothing in the chunk because the innner iterator is
                    // exhausted, or we get a send error (implying the receiver has been
                    // dropped), then exit the thread's main loop.
                    if chunk.is_empty() || sender.send(chunk).is_err() {
                        break;
                    }
```

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 15]
```diff
@@ -0,0 +1,393 @@
...
+                }
+                // TODO - Logging?
+                let _send_result = sender.send(None);
```
Comment (tfenne) [inline src/iterators/chunked_read_ahead_iterator.rs]: Again, why not just `drop(sender)` here?

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 16]
```diff
@@ -0,0 +1,393 @@
...
+            })
+            .expect("failed to spawn chunked read ahead thread");
+
+        // Store the necessary objects on ``Self``
+        Self {
+            current_chunk: Vec::new().into_iter(),
+            receiver: Some(receiver),
+            join_handle: Some(join_handle),
```
Comment (tfenne) [inline src/iterators/chunked_read_ahead_iterator.rs]: Please use the same order as in the struct declaration.

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 17]
```diff
@@ -0,0 +1,393 @@
...
+    }
+}
+
+impl<T> Drop for ChunkedReadAheadIterator<T>
+where
+    T: Send + 'static,
+{
+    fn drop(&mut self) {
+        // Make sure our reciever is dropped so our spawned thread shuts down.
+        self.receiver = None;
```
Comment (tfenne) [inline src/iterators/chunked_read_ahead_iterator.rs]: I feel like the following is clearer in intent
```suggestion
        drop(self.receiver.take());
```

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 18]
```diff
@@ -0,0 +1,368 @@
+use std::any::Any;
+use std::panic::{catch_unwind, resume_unwind, AssertUnwindSafe};
+use std::sync::mpsc::{sync_channel, Receiver};
+use std::thread;
+use std::vec::IntoIter;
+
+// type aliased to get clippy to not think this is too complex
+type PanicUnwindErr = Box<dyn Any + Send>;
+/// Iterator extension that spawns an additional thread to read-ahead in the iterator. Sends
```
Comment (tfenne) [inline src/iterators/chunked_read_ahead_iterator.rs]:
```suggestion
type PanicUnwindErr = Box<dyn Any + Send>;

/// Iterator extension that spawns an additional thread to read-ahead in the iterator. Sends
```

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 19]
```diff
@@ -0,0 +1,368 @@
...
+    {
+        assert_ne!(chunk_size, 0, "Chunk size cannot be zero!");
+        assert_ne!(num_chunk_buffer_size, 0, "Number of buffered chunks cannot be zero!");
+
+        // Create a channel over which we can send our chunks of ``T``
+        let (sender, receiver) = sync_channel(num_chunk_buffer_size);
+
+        // Create our spawned thread, holding on to the resulting handle for downstream error
+        // management.
+        thread::Builder::new()
```
Comment (tfenne) [inline src/iterators/chunked_read_ahead_iterator.rs]: It feels weird not to store the join handle, but I guess we don't need it?

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 20]
```diff
@@ -0,0 +1,368 @@
...
+                    let mut chunk = Vec::with_capacity(chunk_size);
+                    for _ in 0..chunk_size {
+                        match catch_unwind(AssertUnwindSafe(|| inner.by_ref().next())) {
+                            Ok(Some(val)) => chunk.push(val),
+                            Ok(None) => break,
+                            Err(e) => {
+                                sender
+                                    .send(Ok(chunk))
+                                    .expect("Error sending final chunk before internal panic");
+                                sender.send(Err(e)).expect("Error sending internal panic result");
```
Comment (tfenne) [inline src/iterators/chunked_read_ahead_iterator.rs]: I don't _think_ we want to `.expect()` here.  I.e. if the receiver has been dropped, the thread should just go quietly into the night.

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 21]
```diff
@@ -0,0 +1,368 @@
...
+                        }
+                    }
+                    if chunk.is_empty() {
+                        break;
+                    }
+                    let send_result = sender.send(Ok(chunk));
+                    if send_result.is_err() {
+                        // TODO - Logging?
+                        break;
+                    }
```
Comment (tfenne) [inline src/iterators/chunked_read_ahead_iterator.rs]:
```suggestion
                    // If there's nothing in the current chunk that signifies the end of
                    // the inner iterator.  Or if the send fails, the receiver has been
                    // dropped.  In either case, we're done here.
                    if chunk.is_empty() || sender.send(Ok(chunk)).is_err() {
                        break;
                    }
```

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 22]
```diff
@@ -0,0 +1,368 @@
...
+    T: Send + 'static,
+{
+    type Item = T;
+
+    fn next(&mut self) -> Option<T> {
+        // Check if our current chunk has anything left in it
+        // If so, just return that result
+        // If not, see documentation on else block
+        if let Some(result) = self.current_chunk.next() {
+            Some(result)
```
Comment (tfenne) [inline src/iterators/chunked_read_ahead_iterator.rs]: It always feels weird to me to unwrap and then re-wrap a Some.  Would it be better to do:


```rust
let next_option = self.current_chunk.next();
if next_option.is_some() {
    next_option
}
else {
    ....
}
```

File: src/iterators/chunked_read_ahead_iterator.rs (hunk 17 above)
Comment (tfenne) [reply src/iterators/chunked_read_ahead_iterator.rs]: And actually, I think the whole custom `Drop` impl can go away now, and as part of that the Receiver can be stored without an Option, right?

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 23]
```diff
@@ -0,0 +1,368 @@
...
+        let mut chunked_iter = test_vec.into_iter().read_ahead(chunk_size, 1); // TODO add into chunked readahead
+        assert_eq!(chunked_iter.next(), None);
+    }
+
+    #[rstest]
+    #[case(1)] // smallest possible
+    #[case(2)]
+    #[case(4)]
+    #[case(8)]
+    #[case(16)] // larger than the inner iterator
```
Comment (tfenne) [inline src/iterators/chunked_read_ahead_iterator.rs]: Comment doesn't apply to this case, right?
Should we also parameterize over the number of chunks?

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 24]
```diff
@@ -0,0 +1,368 @@
...
+        drop(chunked_iter);
+        let mut test_iter = vec![0usize, 1, 2, 3, 4, 5, 6, 7, 8, 9].into_iter();
+        for i in 0..20 {
+            let v = test_iter.next();
+            if i < 10 {
+                assert_eq!(v, Some(i));
+            } else {
+                assert_eq!(v, None);
+            }
+        }
```
Comment (tfenne) [inline src/iterators/chunked_read_ahead_iterator.rs]: What does the second half of this test do?

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 25]
```diff
@@ -0,0 +1,368 @@
...
+            if i.is_none() {
+                assert!(j.is_none());
+                break;
+            }
+        }
+    }
+
+    #[test]
+    fn test_read_past_end() {
+        let mut test_iter = vec![0usize, 1, 2, 3, 4, 5, 6, 7, 8, 9].into_iter();
```
Comment (tfenne) [inline src/iterators/chunked_read_ahead_iterator.rs]: This needs to be converted into a read_ahead() right?

**Review by tfenne — CHANGES_REQUESTED** [review]

File: src/iterators/chunked_read_ahead_iterator.rs (hunk 24 above)
Comment (NatPRoach) [reply src/iterators/chunked_read_ahead_iterator.rs]: ... This test appears to be a pair of copy paste error that are not doing what its supposed to. Let me fix.

File: src/iterators/chunked_read_ahead_iterator.rs (hunk 19 above)
Comment (NatPRoach) [reply src/iterators/chunked_read_ahead_iterator.rs]: That was my thought. We can store it if you want for future proofing but I don't think it's necessary anymore.

File: src/iterators/chunked_read_ahead_iterator.rs (hunk 5 above)
Comment (NatPRoach) [reply src/iterators/chunked_read_ahead_iterator.rs]: That makes sense. Yeah I wasn't sure what we wanted the import structure to look like.

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 26]
```diff
@@ -0,0 +1,386 @@
...
+// type aliased to get clippy to not think this is too complex
+type PanicUnwindErr = Box<dyn Any + Send>;
+
+/// Iterator extension that spawns an additional thread to read-ahead in the iterator. Sends
+/// results back to this object via a channel and returns them in the same manner as a normal
+/// iterator. This is useful in the context in which the reading of an iterator (or iterators) is
+/// time consuming (e.g. reading from a compressed FASTQ file) and the main thread is bottlenecked
+/// by the speed of the underlying iterator.
+/// 
+/// To use on a struct s that implements ``IntoIter``, it is as simple as:
```
Comment (nh13) [inline src/iterators/chunked_read_ahead_iterator.rs]:
```suggestion
/// To use on a struct that implements ``IntoIter``, it is as simple as:
```

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 27]
```diff
@@ -0,0 +1,386 @@
...
+/// 
+/// let mut chunked_iter = v.into_iter().read_ahead(chunk_size, buffer_size);
+/// assert_eq!(chunked_iter.next(), Some(0));
+/// assert_eq!(chunked_iter.next(), Some(1));
+/// assert_eq!(chunked_iter.next(), Some(2));
+/// assert_eq!(chunked_iter.next(), Some(3));
+/// assert_eq!(chunked_iter.next(), Some(4));
+/// assert_eq!(chunked_iter.next(), Some(5));
+/// assert_eq!(chunked_iter.next(), Some(6));
+/// assert_eq!(chunked_iter.next(), Some(7));
```
Comment (nh13) [inline src/iterators/chunked_read_ahead_iterator.rs]: add in the final next that returns None?

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 28]
```diff
@@ -0,0 +1,386 @@
...
+/// ``` 
+/// Where ``chunk_size`` is the number of elements in the iter to include per send / recieve over
+/// the underlying channel, and ``buffer_size`` is the maximum number of chunks to keep on the
+/// channel at any given time (will block the thread until the space is freed up.).
```
Comment (nh13) [inline src/iterators/chunked_read_ahead_iterator.rs]:
```suggestion
/// channel at any given time (will block the thread until the space is freed up).
```

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 29]
```diff
@@ -0,0 +1,389 @@
...
+///
+/// If your struct does not implement ``IntoIter``, you can either `impl`
+/// ``IntoChunkedReadAheadIterator`` manually or `impl` ``IntoIter`` manually and use the auto
+/// implementation from this module by importing ``IntoChunkedReadAheadIterator``.
+///
+/// The chunked iterator can panic in the following circumstances:
+///     - panics if the underlying iterator panics after the same number of ``next()`` calls.
+pub struct ChunkedReadAheadIterator<T: Send + 'static> {
+    /// The recieving object that recieves chunks of ``T``. TODO - make this a Vec<T> when adding
```
Comment (nh13) [inline src/iterators/chunked_read_ahead_iterator.rs]: TODO still relevant?

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 30]
```diff
@@ -0,0 +1,389 @@
...
+    current_chunk: IntoIter<T>,
+}
+
+impl<T> ChunkedReadAheadIterator<T>
+where
+    T: Send + 'static,
+{
+    /// Creates a new ``Self`` from an existing iterator. Takes two parameters that
+    /// control the number of items stored in the read-ahead buffer.  `chunk_size`
+    /// refers to how many items are transferred at a time from read-ahead thread and
```
Comment (nh13) [inline src/iterators/chunked_read_ahead_iterator.rs]:
```suggestion
    /// refers to how many items are transferred at a time from the read-ahead thread and
```

File: src/iterators/chunked_read_ahead_iterator.rs (hunk 19 above)
Comment (nh13) [reply src/iterators/chunked_read_ahead_iterator.rs]: What if there's some other panic that isn't caught by the `catch_unwind`, for example, we run out of memory and the chunk allocation fails, or a panic occurs when next is called on the inner iterator??  Do we not want to return that at the end of the thread?  So we'd need to join to get that error?

File: src/iterators/chunked_read_ahead_iterator.rs (hunk 3 above)
Comment (nh13) [reply src/iterators/chunked_read_ahead_iterator.rs]: Just create a struct that implements iterator and panics when it is dropped?  Alternatively, just implement the example of reading lines and failing to close (can you fail if you close a second time?)

**Review by nh13 — APPROVED** [review]
Minor changes reviewing the 2nd reviewed version

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 31]
```diff
@@ -0,0 +1,442 @@
...
+use std::thread::{self, JoinHandle};
+use std::vec::IntoIter;
+
+// type aliased to get clippy to not think this is too complex
+type PanicUnwindErr = Box<dyn Any + Send>;
+
+/// Iterator extension that spawns an additional thread to read-ahead in the iterator. Sends
+/// results back to this object via a channel and returns them in the same manner as a normal
+/// iterator. This is useful in the context in which the reading of an iterator (or iterators) is
+/// time consuming (e.g. reading from a compressed FASTQ file) and the main thread is bottlenecked
```
Comment (tfenne) [inline src/iterators/chunked_read_ahead_iterator.rs]:
```suggestion
/// iterator. This is useful in the context in which the reading of an iterator (or iterators) is
/// time consuming or computationally expensive and the consuming thread is bottlenecked
```

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 32]
```diff
@@ -0,0 +1,442 @@
...
+/// assert_eq!(chunked_iter.next(), Some(2));
+/// assert_eq!(chunked_iter.next(), Some(3));
+/// assert_eq!(chunked_iter.next(), Some(4));
+/// assert_eq!(chunked_iter.next(), Some(5));
+/// assert_eq!(chunked_iter.next(), Some(6));
+/// assert_eq!(chunked_iter.next(), Some(7));
+/// assert_eq!(chunked_iter.next(), None);
+/// ```
+/// Where ``chunk_size`` is the number of elements in the iter to include per send / recieve over
+/// the underlying channel, and ``buffer_size`` is the maximum number of chunks to keep on the
```
Comment (tfenne) [inline src/iterators/chunked_read_ahead_iterator.rs]: I'd still prefer to call `buffer_size` something more like `chunk_count` or `chunk_queue_size`, because I think `buffer_size` is more easily misinterpreted if one isn't paying attention.

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 33]
```diff
@@ -0,0 +1,442 @@
...
+// type aliased to get clippy to not think this is too complex
+type PanicUnwindErr = Box<dyn Any + Send>;
+
+/// Iterator extension that spawns an additional thread to read-ahead in the iterator. Sends
+/// results back to this object via a channel and returns them in the same manner as a normal
+/// iterator. This is useful in the context in which the reading of an iterator (or iterators) is
+/// time consuming (e.g. reading from a compressed FASTQ file) and the main thread is bottlenecked
+/// by the speed of the underlying iterator.
+///
+/// To use on a struct that implements ``IntoIter``, it is as simple as:
```
Comment (tfenne) [inline src/iterators/chunked_read_ahead_iterator.rs]: I'm confused by all the references to `IntoIter`.  a) `IntoIter` is a struct returned by `Vec::into_iter()`, not a trait (I think you mean `IntoIterator`), but in general there is no requirement to implement that.  You just need an iterator whose Item type is `Send + 'static`.

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 34]
```diff
@@ -0,0 +1,442 @@
...
+        // If so, just return that result
+        // If not, see documentation on else block
+        let next_option = self.current_chunk.next();
+        if next_option.is_some() {
+            next_option
+        } else {
+            // Current chunk didn't have anything left in it, so
+            // Try to grab a new chunk, and panic if there are no chunks left (note that
+            // ``recv`` is blocking, so this will only return an error if the sender has been
+            // dropped and there are no more elements in the channel.)
```
Comment (tfenne) [inline src/iterators/chunked_read_ahead_iterator.rs]: I don't think "panic if there are no chunks left" is accurate any more, right?

File: src/iterators/chunked_read_ahead_iterator.rs [hunk 35]
```diff
@@ -0,0 +1,442 @@
...
+            // if the new chunk is an Err, raise it to the main thread.
+            match res_r {
+                Ok(next_chunk) => {
+                    self.current_chunk = next_chunk.into_iter();
+                    self.current_chunk.next()
+                }
+                Err(e) => {
+                    resume_unwind(e);
+                }
+            }
```
Comment (tfenne) [inline src/iterators/chunked_read_ahead_iterator.rs]: I still find this whole block really hard to follow. 

E.g. why do `let res_r = if let ...` when the `else` branch returns from the function?

I know if/let is common in rust, but when we have lots of layers of results/options going on I think case/match might be clearer.  

At a minimum I would suggest:

```suggestion
            if let Ok(chunk_or_panic) = self.receiver.recv() {
                // If the new chunk is present and Ok, convert it to an iterator, store it on ``self``,
                // and return its next value ( shutting down our reciever if the next value is None).
                // if the new chunk is an Err, raise it to the main thread.
                match chunk_or_panic {
                    Ok(next_chunk) => {
                        self.current_chunk = next_chunk.into_iter();
                        self.current_chunk.next()
                    }
                    Err(e) => {
                        resume_unwind(e);
                    }
                }
            } else {
                // join handle is not ``Copy`` or ``Clone`` and we need ownership of it to be able
                // to join on it, hence the optional field and taking it off the iterator struct.
                if let Some(join_handle) = self.join_handle.take() {
                    if let Err(e) = join_handle.join() {
                        resume_unwind(e)
                    }
                }
                return None;
            };
```

**Review by tfenne — APPROVED** [review]
